README.md
# Mr Clippy

## Installation

clippy/CHANGELOG.md
 # Change Log

All notable changes to the "clippy" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

bing_search.js
const bing_search = (() => {
    const bingUrl = "https://bing.com/search?q="
    const msSupportSites = [
        "support.microsoft.com",
        "support.microsoft.com/en-us/microsoft-365",
        "support.microsoft.com/en-us/outlook",
        "support.microsoft.com/en-us/word",
        "support.microsoft.com/en-us/excel",
        "support.microsoft.com/en-us/powerpoint",
        "support.microsoft.com/en-us/teams",
        "docs.microsoft.com/en-us/windows"
    ]
    let roundRobinPointer = 0

    function generate_link(input) {
        let queryPrefix = `site:${msSupportSites[roundRobinPointer]} `

        roundRobinPointer++

        if (roundRobinPointer > msSupportSites.length - 1)
            roundRobinPointer = 0

        return bingUrl + encodeURI(queryPrefix + input)
    }

    return {
        generate_link
    }
})();

module.exports = {
    generate_link: bing_search.generate_link
}

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import * as path from 'path';
import * as bing_search from './bing-search';
import { serialize } from 'v8';

const bingSearch = bing_search.bingSearch

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
	let clippyPanel:any = null

	context.subscriptions.push(
		vscode.window.onDidChangeTextEditorSelection((event) => {
			let text = event.textEditor.document.getText(event.selections[0])

			if (text) {
				if (clippyPanel)
					clippyPanel.dispose()

				let helpText = clippySays(text)

				console.log(helpText)

				clippyPanel = clippy(context, helpText.say, helpText.linkText)
			}
		})
	)
}

function createHelpText(match:Object) {
	// @ts-ignore
	return `It looks like you're trying to use ${match.category}!`;
}

function resolveCategory(matchTerm:string = "") {
	const termMap:Object = {
		"loops": ["for", "while"],
		"statement": ["switch", "if", "else"],
		"html": ["html", "body", "div"],
		"variables": ["const", "let", "var"],
		"brackets": ["{", "}", "[", "]"],
		"functions": ["function", "(", ")"],
		"classes": ["class"],
		"styles": ["style"]
	}

	for (let category in termMap) {
		// @ts-ignore
		if (termMap[category].indexOf(matchTerm) !== -1) {
			return {matchTerm, category}
		}
	}

	return {matchTerm, category: ""}
}

function clippySays(text:string) {
	const words:Array<string> = text.split(" ")
				
	// Loop over the selection in reverse and find the first match
	let match:Object = {}

	for (let i:number = words.length - 1; i >= 0; i--) {
		match = resolveCategory(words[i].toLowerCase())
		
		// @ts-ignore
		if (match.category !== "")
			break;
	}

	// @ts-ignore
	let linkText = bingSearch.generate_link(match.category);

	// @ts-ignore
	if (match.category === "")
		return {say: "Unable to find any results", linkText};

	// @ts-ignore
	let say = createHelpText(match)

	return {say, linkText}
}

function clippy(context: any, text: string, link: string) {
	// Create and show a new webview
	const panel = vscode.window.createWebviewPanel(
	  'clippy', // Identifies the type of the webview. Used internally
	  'Clippy', // Title of the panel displayed to the user
	  {
		  viewColumn: vscode.ViewColumn.Beside, // Editor column to show the new webview panel in.
		  preserveFocus: true
	  },
	  {
		enableScripts: true,

		localResourceRoots: [
			context.extensionUri
		],
		
	  }
	);

	const clippy_js_on_disk = vscode.Uri.file(
		path.join(context.extensionPath, "src/clippy/build/clippy.min.js")
	);
	const clippy_css_on_disk = vscode.Uri.file(
		path.join(context.extensionPath, "src/clippy/build/clippy.css")
	)

	//@ts-ignore
	const clippy_js = panel.webview.asWebviewUri(clippy_js_on_disk)
	//@ts-ignore
	const clippy_css = panel.webview.asWebviewUri(clippy_css_on_disk)

	// And set its HTML content
	panel.webview.html = getWebviewContent(clippy_js, clippy_css, text, link);

	return panel
}

function getWebviewContent(
	clippy_js:any, 
	clippy_css:any, 
	text:string,
	link:string
) {
	return   `<!DOCTYPE html>
		<html lang="en">
		<head>
			<meta charset="UTF-8">
			<meta name="viewport" content="width=device-width, initial-scale=1.0">
			<title>Clippy</title>
			<link rel="stylesheet" type="text/css" href="${clippy_css}" media="all">
		</head>
		<body>
			<script src="https://code.jquery.com/jquery-1.12.4.min.js" integrity="sha256-ZosEbRLbNQzLpnKIkEdrPv7lOy9C27hHQ+Xp8a4MxAQ=" crossorigin="anonymous"></script>
			<script src="${clippy_js}"></script>
			<!-- Init script -->
			<script type="text/javascript">
				let input = document.querySelector('#name');

				clippy.load('Clippy', function(agent) {
					agent.animate();
					speak(agent, "${text}", "${link}");
				});

				const speak = (agent, speech, link) => {
					agent.speak("Searching...");
					agent.play('Searching', 5000, () => {
						agent.speak(speech, true);

						let searchLink = document.createElement('a')
						searchLink.href = link
						searchLink.innerHTML = "get help online"
						document.getElementsByClassName('clippy-balloon')[0].appendChild(searchLink)

						// Hack it sometimes not setting the height of the text box correctly.
						setTimeout(() => {
							document.querySelector('.clippy-content').style.height = 'auto';
						}, 1000)
					});	
				}
			</script>
		</body>
		</html>`
}

// this method is called when your extension is deactivated
export function deactivate() {}


out/bing-search.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bingSearch = void 0;
exports.bingSearch = (() => {
    const bingUrl = "https://bing.com/search?q=";
    const msSupportSites = [
        "support.microsoft.com",
        "support.microsoft.com/en-us/microsoft-365",
        "support.microsoft.com/en-us/outlook",
        "support.microsoft.com/en-us/word",
        "support.microsoft.com/en-us/excel",
        "support.microsoft.com/en-us/powerpoint",
        "support.microsoft.com/en-us/teams",
        "docs.microsoft.com/en-us/windows"
    ];
    let roundRobinPointer = 0;
    function generate_link(input = "") {
        let queryPrefix = `site:${msSupportSites[roundRobinPointer]} `;
        roundRobinPointer++;
        if (roundRobinPointer > msSupportSites.length - 1)
            roundRobinPointer = 0;
        return bingUrl + encodeURI(queryPrefix + input);
    }
    return {
        generate_link
    };
})();
//# sourceMappingURL=bing-search.js.map

src/bing-search.ts
export const bingSearch = (() => {
    const bingUrl = "https://bing.com/search?q="
    const msSupportSites = [
        "support.microsoft.com",
        "support.microsoft.com/en-us/microsoft-365",
        "support.microsoft.com/en-us/outlook",
        "support.microsoft.com/en-us/word",
        "support.microsoft.com/en-us/excel",
        "support.microsoft.com/en-us/powerpoint",
        "support.microsoft.com/en-us/teams",
        "docs.microsoft.com/en-us/windows"
    ]
    let roundRobinPointer = 0

    function generate_link(input:string = "") {
        let queryPrefix = `site:${msSupportSites[roundRobinPointer]} `

        roundRobinPointer++

        if (roundRobinPointer > msSupportSites.length - 1)
            roundRobinPointer = 0

        return bingUrl + encodeURI(queryPrefix + input)
    }

    return {
        generate_link
    }
})();


out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require("vscode");
const path = require("path");
const bing_search = require("./bing-search");
const bingSearch = bing_search.bingSearch;
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
function activate(context) {
    let clippyPanel = null;
    context.subscriptions.push(vscode.window.onDidChangeTextEditorSelection((event) => {
        let text = event.textEditor.document.getText(event.selections[0]);
        if (text) {
            if (clippyPanel)
                clippyPanel.dispose();
            let helpText = clippySays(text);
            console.log(helpText);
            clippyPanel = clippy(context, helpText.say, helpText.linkText);
        }
    }));
}
exports.activate = activate;
function createHelpText(match) {
    // @ts-ignore
    return `It looks like you're trying to use ${match.category}!`;
}
function resolveCategory(matchTerm = "") {
    const termMap = {
        "loops": ["for", "while"],
        "statement": ["switch", "if", "else"],
        "html": ["html", "body", "div"],
        "variables": ["const", "let", "var"],
        "brackets": ["{", "}", "[", "]"],
        "functions": ["function", "(", ")"],
        "classes": ["class"],
        "styles": ["style"]
    };
    for (let category in termMap) {
        // @ts-ignore
        if (termMap[category].indexOf(matchTerm) !== -1) {
            return { matchTerm, category };
        }
    }
    return { matchTerm, category: "" };
}
function clippySays(text) {
    const words = text.split(" ");
    // Loop over the selection in reverse and find the first match
    let match = {};
    for (let i = words.length - 1; i >= 0; i--) {
        match = resolveCategory(words[i].toLowerCase());
        // @ts-ignore
        if (match.category !== "")
            break;
    }
    // @ts-ignore
    let linkText = bingSearch.generate_link(match.category);
    // @ts-ignore
    if (match.category === "")
        return { say: "Unable to find any results", linkText };
    // @ts-ignore
    let say = createHelpText(match);
    return { say, linkText };
}
function clippy(context, text, link) {
    // Create and show a new webview
    const panel = vscode.window.createWebviewPanel('clippy', // Identifies the type of the webview. Used internally
    'Clippy', // Title of the panel displayed to the user
    {
        viewColumn: vscode.ViewColumn.Beside,
        preserveFocus: true
    }, {
        enableScripts: true,
        localResourceRoots: [
            context.extensionUri
        ],
    });
    const clippy_js_on_disk = vscode.Uri.file(path.join(context.extensionPath, "src/clippy/build/clippy.min.js"));
    const clippy_css_on_disk = vscode.Uri.file(path.join(context.extensionPath, "src/clippy/build/clippy.css"));
    //@ts-ignore
    const clippy_js = panel.webview.asWebviewUri(clippy_js_on_disk);
    //@ts-ignore
    const clippy_css = panel.webview.asWebviewUri(clippy_css_on_disk);
    // And set its HTML content
    panel.webview.html = getWebviewContent(clippy_js, clippy_css, text, link);
    return panel;
}
function getWebviewContent(clippy_js, clippy_css, text, link) {
    return `<!DOCTYPE html>
		<html lang="en">
		<head>
			<meta charset="UTF-8">
			<meta name="viewport" content="width=device-width, initial-scale=1.0">
			<title>Clippy</title>
			<link rel="stylesheet" type="text/css" href="${clippy_css}" media="all">
		</head>
		<body>
			<script src="https://code.jquery.com/jquery-1.12.4.min.js" integrity="sha256-ZosEbRLbNQzLpnKIkEdrPv7lOy9C27hHQ+Xp8a4MxAQ=" crossorigin="anonymous"></script>
			<script src="${clippy_js}"></script>
			<!-- Init script -->
			<script type="text/javascript">
				let input = document.querySelector('#name');

				clippy.load('Clippy', function(agent) {
					agent.animate();
					speak(agent, "${text}", "${link}");
				});

				const speak = (agent, speech, link) => {
					agent.speak("Searching...");
					agent.play('Searching', 5000, () => {
						agent.speak(speech, true);

						let searchLink = document.createElement('a')
						searchLink.href = link
						searchLink.innerHTML = "get help online"
						document.getElementsByClassName('clippy-balloon')[0].appendChild(searchLink)

						// Hack it sometimes not setting the height of the text box correctly.
						setTimeout(() => {
							document.querySelector('.clippy-content').style.height = 'auto';
						}, 1000)
					});	
				}
			</script>
		</body>
		</html>`;
}
// this method is called when your extension is deactivated
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map